File: python/해시/폰켓몬.py

""" 레벨 1. 폰켓몬 """

def solution(nums):
    
    mons = set()
    for n in nums:
        mons.add(n)

    count = len(nums) // 2
    return count if count <= len(mons) else len(mons)

print(solution([3,1,2,3]))          # 2
print(solution([3,3,3,2,2,4]))      # 3
print(solution([3,3,3,2,2,2]))      # 2

File: Java/src/자료구조/폰켓몬.java
package 자료구조;

// import java.util.HashMap;
import java.util.HashSet;

class Solution {

    public static void main(String[] args) {

        int[] s1 = {3, 1, 2, 3};
        System.out.println(solution(s1));

        int[] s2 = {3,3,3,2,2,4};
        System.out.println(solution(s2));

        int[] s3 = {3,3,3,2,2,2};
        System.out.println(solution(s3));

    }

    public static int solution(int[] nums) {

        HashSet<Integer> set = new HashSet<>();

        for (int num : nums) {
            set.add(num);
        }

        return Math.min(nums.length / 2, set.size());
    }
}